### PR TITLE
feat/ Write Aura Protocol Integrations Smart Contracts

### PR DESCRIPTION
**Pull request by 0xEinCodes**
## Core changes:

This PR encompasses the integration of AURA into Sommelier `cellars` via the use of an `AuraERC4626Adaptor` which has some extra `strategist functions` atop of the 'general' `ERC4626Adaptor` functionality (essentially a copy of the `CellarAdaptor.sol` but with some reformatting). 

> Initially the design of this adaptor encompassed use the `CellarAdaptor.sol` and a contract called `AuraExtrasAdaptor.sol` but they were combined into `AuraERC4626Adaptor.sol` inheriting the `ERC4626Adaptor.sol` after some discussion.

> Currently, this PR is in the discussion and design phase.

### At a high level:
Tests are to be written to assess the functionality of using a `ERC4626Adaptor` with AURA `BaseRewardPool4626.sol` for each BPT accepted in the AURA protocol. An ERC4626 adaptor position is also being considered for integration into the AuraBal Compounder Vault. 

Although, custom integration is needed with the compounder and vote proxy smart contracts for staking auraBAL, and participating in AURA's voting schema by locking AURA, respectively. These last features would be tackled in the future, and are outside of the scope of this PR.

## TODO

- [x] Assess Aura docs && github && deployed contracts for base functions
- [x] " for specialized functionality (staking auraBAL, locking Aura and voting, etc.)
- [x] Outline the function call path that is expected when going through basic deposit and withdraw of BPTs, and staking of aura-BPTs into AURA protocol, and claiming of rewards.
- [x] Discuss the basic design paths with each contract (cellar for basic functionality, auraBALAdaptor, ~lockAuraAdaptor~) in AURA telegram chat w/ PR
- [x] ~Write the outline of the auraBALAdaptor contract~
- [x] ~Write the outline of the lockAuraAdaptor contract~
- [x] Write the basic function tests (or stub outs at least) for using cellar adaptor w/ AURA pools. Use this for discussion.
- [x] Finish basic function tests
- [x] Write tests for custom adaptors
- [x] PR Review

## Live Discussion Topics (discussed during the PR stage)
- This is to be rebased once `dev/macro-audit-10` is created, where the audit for `dev/macro-audit-9` is being merged into `main` this coming week. `dev/macro-audit-10` is scheduled for end of September 2023.
- Within the docs it states that: Aura is price agnostic and thus does not require oracles.  Flash loans or any other type of price manipulation or front running do not affect the Aura system. I need to look into what this implies.

## References

1. [Aura docs](https://docs.aura.finance/)
2. [Mainnet Deployed Example of Aura Pool](https://etherscan.deth.net/address/0x032B676d5D55e8ECbAe88ebEE0AA10fB5f72F6CB)
3. [Mainned Deployed AuraBal Compounder Vault](https://etherscan.deth.net/address/0xfAA2eD111B4F580fCb85C48E6DC6782Dc5FCD7a6)
4. [Aura BaseRewardPool Github](https://github.com/aurafinance/convex-platform/blob/816cbfd551a80bb4768f9168144dadbd3e35bd13/contracts/contracts/BaseRewardPool.sol#L238C1-L239C1)
5. [Aura BaseRewardPool4626](https://github.com/aurafinance/convex-platform/blob/lite/contracts/contracts/BaseRewardPool4626.sol)
6. [Aura Main Contracts (outside of base reward pools)](https://github.com/aurafinance/aura-contracts/tree/main/contracts)
7. [Convex docs](https://docs.convexfinance.com/convexfinanceintegration/booster)
8. [Convex github (booster specifically)](https://github.com/convex-eth/platform/blob/main/contracts/contracts/Booster.sol)
9. [Diff check btw AURA and CONVEX at earlier commit](https://github.com/aurafinance/convex-platform/pull/23/files?file-filters%5B%5D=.sol&show-deleted-files=true&show-viewed-files=true#diff-ba284ba6baa92343fd84323538d74e2b78311d6a465f2ef97f6f32b3e70426b2)
10. [Hidden Hand](https://hiddenhand.finance/aura)
---

<details markdown='1'><summary> ℹ️ Extra Notes </summary>

## Aura Architecture Notes

tl;dr - It looks like AURA is an adjusted fork of Convex. Some of the features they have included now are: renamed contracts && introduction of `BaseRewardPool4626`, and non-ERC4626 auraBALVaults && voteLockAuraVaults I believe. We mainly will be working in the `BaseRewardPool4626` contracts. 

- `Booster` contracts, from Convex codebase, are used within the `BaseRewardPool4626` contracts to stake the respective `assetToken` and `mint` the 'CVX' style, in this case aura-BPT, to the receiver. 
- To access `auraBAL`, we either deposit 8020 BAL ETH BPT, or buy auraBAL (via Uniswap or we'd need Balancer Swap integration). 
- Once we have `auraBAL` we can stake it within the auraBALVault contract(s), and earn.
- Finally `voteLockAura` would require integration with the auraVault contract. This is last to be developed because I'm not sure if Sommelier will embark on strategies that convert fees earned to voteLockedAura. 
- There are 'bribes' from hidden hand && reward boosts (in full) that Sommelier would be leaving on the table if they did not convert fees to voteLockedAura of their own.

- Leaning towards using a cellarAdaptor with Aura BaseRewardPool4626.sol. It can be used to enter and exit, and keep accounting of the base assets (BPTs) involved with the cellar position.
- Rewards are TBD --> it seems they have to be claimed though upon withdrawal. So the cellar would not know about the rewards until claiming the actual tokens.
- So ERC20Adaptors are needed to keep track of the rewards being claimed is all.
- Special Aura adaptors would then be needed I think to do voteLock AURA and stake auraBAL. 

Tested claiming rewards with Aura-rETH-WETH Pool: 
rETH_WETH aura BPT vAPR
- The main intent of reward claiming is to ensure we can claim all $BAL and all $AURA allocated to cellar positions.

</details>

<!--- example - Design decisions made in PR>
<!--- example - Remember running `forge fmt` before committing the PR>
<!--- example - Reference for how to run relevant tests i.e. forge test --match-contract X --fork-url $RPC_URL -vvv>

### CODE REVIEW

File: src/modules/adaptors/Aura/AuraERC4626Adaptor.sol [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    //============================================ Interface Helper Functions ===========================================
+
+    //============================== Interface Details ==============================
+    // It is unlikely, but AURA pool interfaces can change between versions.
+    // To account for this, internal functions will be used in case it is needed to
+    // implement new functionality.
+    //===============================================================================
+
+    function _getRewards(IBaseRewardPool _auraPool, bool _claimExtras) internal virtual {
```
Comment (crispymangoes) [inline src/modules/adaptors/Aura/AuraERC4626Adaptor.sol:75]: Oh did you separate this into a virtual internal function because they might change the `getReward` function in the future? Kinda like you did with Fraxlend adaptors?

File: src/modules/adaptors/Sommelier/ERC4626Adaptor.sol [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+/**
+ * @title Generic ERC4626 Vault Adaptor (basically a copy of Cellar Adaptor w/ virtual function sigs to enable inheritance and overriding).
+ * @notice Allows Cellars to interact with other ERC4626 contracts.
+ * @author crispymangoes, 0xEinCodes
+ * @dev this is a WIP and not audited yet.
+ * NOTE: `CellarAdaptor.sol` was used, and is still used for nested Cellar positions for Sommelier Cellars (ERC4626 Vaults). Integrations into Aura, among other ERC4626 contracts outside of Sommelier have led to using a separate ERC4626Vault.sol file that allows more flexibility via inheritance.
+ * NOTE: `Cellar` is still used throughout this contract at times because it is a contract made for the Sommelier protocol (has extra pricing and accounting mechanics, etc.)
+ */
+contract ERC4626Adaptor is BaseAdaptor {
```
Comment (crispymangoes) [inline src/modules/adaptors/Sommelier/ERC4626Adaptor.sol:15]: I think this contract should be moved to be with the `ERC20Adaptor.sol`, since its not a somm specific contract, but interested to hear your thoughts

File: src/modules/adaptors/Sommelier/ERC4626Adaptor.sol [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    //============================================ Strategist Functions ===========================================
+    /**
+     * @notice Allows strategists to deposit into ERC4626 positions.
+     * @dev Uses `_maxAvailable` helper function, see BaseAdaptor.sol
+     * @param erc4626Vault the ERC4626 to deposit `assets` into
+     * @param assets the amount of assets to deposit into `cellar`
+     */
+    function depositToCellar(ERC4626 erc4626Vault, uint256 assets) public {
```
Comment (crispymangoes) [inline src/modules/adaptors/Sommelier/ERC4626Adaptor.sol:139]: Can you change the strategist withdraw and deposit functions to be `depositToVault` and `withdrawFromVault`

File: test/testAdaptors/AuraERC4626Adaptor.t.sol [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        );
+    }
+
+    // same as redeem test but with a long time within the vault. Shows that rewards are not the vault asset (BPT) but other tokens.
+    function testFastForward(uint256 assets) external {
+        assets = bound(assets, 0.1e18, 1_000_000_000e18);
+        deal(address(rETH_wETH_BPT), address(this), assets);
+        cellar.deposit(assets, address(this)); // bpts sent to aura pool as seen in deposit() tests
+
+        vm.warp(block.timestamp + 100 days);
```
Comment (crispymangoes) [inline test/testAdaptors/AuraERC4626Adaptor.t.sol:203]: I recently discovered you can use `skip(100 days)` instead if you want.

File: src/modules/adaptors/Aura/AuraERC4626Adaptor.sol (hunk 1 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Aura/AuraERC4626Adaptor.sol:75]: Yes, that was the intention behind having a helper function. I did not discuss the % likelihood of new pool versions with the Aura team, but you never know so I thought it was a fair add-on. I can remove it and make it directly part of the `getRewards()` function though. wdyt?

File: src/modules/adaptors/Sommelier/ERC4626Adaptor.sol (hunk 2 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Sommelier/ERC4626Adaptor.sol:15]: When you say "be with the `ERC20Adaptor.sol`." Do you mean it's file location? If so, yes I agree.

File: src/modules/adaptors/Sommelier/ERC4626Adaptor.sol (hunk 3 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Sommelier/ERC4626Adaptor.sol:139]: Done!

File: test/testAdaptors/AuraERC4626Adaptor.t.sol (hunk 4 above)
Comment (0xEinCodes) [reply test/testAdaptors/AuraERC4626Adaptor.t.sol:203]: That seems cleaner. Just tried it and it works!

File: src/modules/adaptors/Sommelier/ERC4626Adaptor.sol (hunk 2 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Sommelier/ERC4626Adaptor.sol]: I've moved the file to where `ERC20Adaptor.sol` is stored within the repo.

**Comment by 0xEinCodes** [comment]
@crispymangoes thanks for reviewing the PR! I've made the changes requested and responded to your comments. Please see `CellarWithERC4626Adaptor.t.sol` and assess if there are parts to be removed/added to make it encompass the testing intent from `ERC4626Adaptor.nc`

File: src/modules/adaptors/Aura/AuraERC4626Adaptor.sol [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        _auraPool.getReward(address(this), _claimExtras); // TODO: confirm that any and all reward tokens associated to this position will be transferred from this external call.
```
Comment (crispymangoes) [inline src/modules/adaptors/Aura/AuraERC4626Adaptor.sol:76]: Was this TODO checked?

File: src/modules/adaptors/ERC4626Adaptor.sol [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+    //============================================ Implement Base Functions ===========================================
+    /**
+     * @notice Cellar must approve ERC4626 position to spend its assets, then deposit into the ERC4626 position.
+     * @param assets the amount of assets to deposit into the ERC4626 position
+     * @param adaptorData adaptor data containining the abi encoded ERC4626
+     * @dev configurationData is NOT used
+     */
+    function deposit(uint256 assets, bytes memory adaptorData, bytes memory) public virtual override {
+        // Deposit assets to `cellar`.
```
Comment (crispymangoes) [inline src/modules/adaptors/ERC4626Adaptor.sol:57]: Change to Deposit assets to `vault`

File: src/modules/adaptors/ERC4626Adaptor.sol [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+pragma solidity 0.8.21;
+
+import { BaseAdaptor, ERC20, SafeTransferLib, Cellar } from "src/modules/adaptors/BaseAdaptor.sol";
+import { ERC4626 } from "@solmate/mixins/ERC4626.sol";
+
+/**
+ * @title Generic ERC4626 Vault Adaptor (basically a copy of Cellar Adaptor w/ virtual function sigs to enable inheritance and overriding).
+ * @notice Allows Cellars to interact with other ERC4626 contracts.
+ * @author crispymangoes, 0xEinCodes
+ * @dev this is a WIP and not audited yet.
```
Comment (crispymangoes) [inline src/modules/adaptors/ERC4626Adaptor.sol:11]: Remove this once review is done

File: test/testAdaptors/AuraERC4626Adaptor.t.sol [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+        price = uint256(IChainlinkAggregator(address(mockRethUsd)).latestAnswer());
+        settings = PriceRouter.AssetSettings(CHAINLINK_DERIVATIVE, address(mockRethUsd));
+        priceRouter.addAsset(rETH, settings, abi.encode(stor), price);
+
+        price = uint256(IChainlinkAggregator(BAL_USD_FEED).latestAnswer());
+        settings = PriceRouter.AssetSettings(CHAINLINK_DERIVATIVE, BAL_USD_FEED);
+        priceRouter.addAsset(BAL, settings, abi.encode(stor), price);
+
+        // TODO: AURA doesn't have an AURA_USD Chainlink Feed. For now, we'll make it a mock price feed with WETH FEED
```
Comment (crispymangoes) [inline test/testAdaptors/AuraERC4626Adaptor.t.sol:78]: But we shouldn't need to price Aura, since strategists can just sell it using a DEC Aggregator adaptor

File: test/testAdaptors/AuraERC4626Adaptor.t.sol [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        price = uint256(IChainlinkAggregator(WETH_USD_FEED).latestAnswer());
+        settings = PriceRouter.AssetSettings(CHAINLINK_DERIVATIVE, WETH_USD_FEED);
+        priceRouter.addAsset(AURA, settings, abi.encode(stor), price);
+
+        // Add rETH_wETH_BPT pricing.
+        uint8[8] memory rateProviderDecimals;
+        address[8] memory rateProviders;
+        ERC20[8] memory underlyings; // TODO: check with CRISPY to get underlying order correct
```
Comment (crispymangoes) [inline test/testAdaptors/AuraERC4626Adaptor.t.sol:86]: Order is right

File: test/testAdaptors/AuraERC4626Adaptor.t.sol [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        vm.warp(block.timestamp + 100 days); // fast forward to accrue rewards
+        mockWethUsd.setMockUpdatedAt(block.timestamp);
+        mockRethUsd.setMockUpdatedAt(block.timestamp);
+
+        uint256 oldBALRewards = BAL.balanceOf(address(cellar));
+        uint256 oldAURARewards = AURA.balanceOf(address(cellar));
+        uint256 oldBPTBalance = rETH_wETH_BPT.balanceOf(address(cellar));
+
+        Cellar.AdaptorCall[] memory data = new Cellar.AdaptorCall[](1);
+        // Withdraw FRAX from FraxLend.
```
Comment (crispymangoes) [inline test/testAdaptors/AuraERC4626Adaptor.t.sol:239]: Old comment

File: test/testAdaptors/AuraERC4626Adaptor.t.sol [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+        bool claimExtras = true;
+        {
+            bytes[] memory adaptorCalls = new bytes[](1);
+            adaptorCalls[0] = _createBytesDataGetRewardsFromAuraPoolERC4626(address(aura_rETH_wETH_BPT), claimExtras);
+            data[0] = Cellar.AdaptorCall({ adaptor: address(auraERC4626Adaptor), callData: adaptorCalls });
+        }
+
+        // Perform callOnAdaptor.
+        cellar.callOnAdaptor(data); // TODO: EIN, check the logs to see what reward tokens are claimed with bool set to false. Then compare to what it is as true.
```
Comment (crispymangoes) [inline test/testAdaptors/AuraERC4626Adaptor.t.sol:248]: Did you do this?

File: test/testAdaptors/ERC4626Adaptor.nc [hunk 12]
```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (crispymangoes) [inline test/testAdaptors/ERC4626Adaptor.nc:1]: This can be deleted

**Review by crispymangoes — APPROVED** [review]
Very minor nit picks. Once done feel free to rebase to dev branch, then merge into dev branch.

File: src/modules/adaptors/Aura/AuraERC4626Adaptor.sol (hunk 5 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Aura/AuraERC4626Adaptor.sol:76]: Done. Double checked it via `console.log` the balances for `BAL`, `AURA`, and `BPTs`, where the first two asset balances should have increased, and the latter should have stayed the same, within the test file. Below are the results. 

The green percentages are the actual normalized (USD) rewards from the test for 100e18 BPT being deposited into Aura. The reddish highlighted percentages are the BAL/AURA split according to the AURA UI on their vAPR. 

There is small discrepancy, but that may be due to other fees I'm missing here. wdyt? I could send this question to Aura too.

image

Also, please note that I did not check for obscure assets that may be part of third party liquidity mining outside of $BAL and $AURA rewards. We discussed how if this strategist function did not claim specific rewards that we could deploy standalone adaptors to do so, and also handle the rewards accordingly (simply claiming them or otherwise).

File: src/modules/adaptors/ERC4626Adaptor.sol [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    //      position can support use withdraws
+    // else:
+    //      position can not support user withdraws
+    //
+    //====================================================================
+
+    /**
+     * @notice Strategist attempted to interact with a erc4626Vault with no position setup for it.
+     */
+    error ERC4626Adaptor__CellarPositionNotUsed(address erc4626Vault);
```
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/adaptors/ERC4626Adaptor.sol]: Done.

File: src/modules/adaptors/ERC4626Adaptor.sol (hunk 6 above)
Comment (0xEinCodes) [reply src/modules/adaptors/ERC4626Adaptor.sol]: Done.

File: src/modules/adaptors/ERC4626Adaptor.sol (hunk 7 above)
Comment (0xEinCodes) [reply src/modules/adaptors/ERC4626Adaptor.sol]: Done.

File: test/testAdaptors/AuraERC4626Adaptor.t.sol (hunk 9 above)
Comment (0xEinCodes) [reply test/testAdaptors/AuraERC4626Adaptor.t.sol]: Awesome, thanks!

File: test/testAdaptors/AuraERC4626Adaptor.t.sol (hunk 10 above)
Comment (0xEinCodes) [reply test/testAdaptors/AuraERC4626Adaptor.t.sol]: Removed.

File: test/testAdaptors/AuraERC4626Adaptor.t.sol (hunk 11 above)
Comment (0xEinCodes) [reply test/testAdaptors/AuraERC4626Adaptor.t.sol]: Yes, I double checked by running console logs on the balances. These console logs are the same that were mentioned in your other comment before within the actual adaptor design.

File: test/testAdaptors/ERC4626Adaptor.nc (hunk 12 above)
Comment (0xEinCodes) [reply test/testAdaptors/ERC4626Adaptor.nc]: Deleted the file.

**Comment by 0xEinCodes** [comment]
Thanks for the PR review @crispymangoes! Please go over the changes that I've made to resolve your change requests and comments. I'm changing the base branch for this PR to `dev/macro-audit-12` branch as well now.

File: test/testAdaptors/AuraERC4626Adaptor.t.sol (hunk 8 above)
Comment (crispymangoes) [reply test/testAdaptors/AuraERC4626Adaptor.t.sol]: Typo meant to type "DEX" like decentralized exchange

Comment (0xEinCodes) [reply test/testAdaptors/AuraERC4626Adaptor.t.sol]: Awesome, thanks for the clarification. I've removed it.

File: src/modules/adaptors/Aura/AuraERC4626Adaptor.sol [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+ * @notice Carries out typical ERC4626Adaptor functionality and allows Cellars to claim rewards from AURA pools
+ * @author crispymangoes, 0xEinCodes
+ * NOTE: Transferrance of aura-wrapped BPT is not alowed as per their contracts: ref - https://etherscan.io/address/0xdd1fe5ad401d4777ce89959b7fa587e569bf125d#code#F1#L254
+ */
+contract AuraERC4626Adaptor is ERC4626Adaptor {
+    using SafeTransferLib for ERC20;
+    using Math for uint256;
+
+    //==================== Adaptor Data Specification ====================
+    // adaptorData = abi.encode(address auraPool) // TODO: CRISPY, the adaptor data comes out as an address even if we specify a different interface such as IERC4626 or ERC4626 right? So this should stay as `address auraPool` or should it be `ERC4626 auraPool`
```
Comment (crispymangoes) [inline src/modules/adaptors/Aura/AuraERC4626Adaptor.sol:21]: So it does not really matter if you encode an address, interface, or even contract type(like ERC20) they are all encoded the same way. So it is totally fine to leave this as an aura pool address, then remove the TODO comment. When ever we are decoding the adaptor data, it will be casted into whatever interface we need it to be. Kinda like how if you encode a `uint8` it encodes it as if it were a `uint256` but when you decode the data, it is type casted back to a `uint8`

**Review by crispymangoes — APPROVED** [review]

File: src/modules/adaptors/Aura/AuraERC4626Adaptor.sol (hunk 14 above)
Comment (0xEinCodes) [reply src/modules/adaptors/Aura/AuraERC4626Adaptor.sol]: Great, thank you for the clarification! I've removed the TODO comments within the `AuraERC4626Adaptor.sol` file.
